modules/EnergyManager/EnergyManagerImpl.cpp

```diff
@@ -0,0 +1,230 @@
+            //     create local offer at evse's marketplace
+
+            offer_tp.start();
+            Offer local_offer(broker->get_local_market());
+            offer_tp.pause();
+
+            // ask broker to trade
+            broker_tp.start();
+            if (broker->trade(local_offer))
+                trade_happend_in_this_round = true;
+            broker_tp.pause();
+        }
+        if (!trade_happend_in_this_round)
+            break;
+    }
+
+    if (max_number_of_trading_rounds <= 0) {
+        EVLOG_error << "Trading: Maximum number of trading rounds reached.";
+    }
+
+    if (globals.debug) {
+        EVLOG_info << fmt::format("\033[1;44m---------------- End energy optimizer ({} rounds, offer {}ms market {}ms "
+                                  "broker {}ms total {}ms) ---------------- \033[1;0m",
+                                  100 - max_number_of_trading_rounds, offer_tp.stop(), market_tp.stop(),
+                                  broker_tp.stop(), optimizer_start.stop());
+    }
+
+    std::vector<types::energy::EnforcedLimits> optimized_values;
+    optimized_values.reserve(brokers.size());
+
+    for (auto& broker : brokers) {
+        auto& local_market = broker->get_local_market();
+        const auto sold_energy = local_market.get_sold_energy();
+
+        if (sold_energy.size() > 0) {
+
+            types::energy::EnforcedLimits l;
+            l.uuid = local_market.energy_flow_request.uuid;
+            l.valid_until =
+                Everest::Date::to_rfc3339(globals.start_time + std::chrono::seconds(config.update_interval * 10));
+
+            l.schedule = sold_energy;
+
+            // select root limit from schedule based on globals.start_time
+            l.limits_root_side = sold_energy[0].limits_to_root;
+
+            for (const auto& s : sold_energy) {
+                const auto schedule_time = Everest::Date::from_rfc3339(s.timestamp);
+                if (globals.start_time < schedule_time) {
+                    // all further schedules will be further into the future
+                    break;
+                } else {
+                    // use this schedule as the starting point
+                    l.limits_root_side = s.limits_to_root;
+                }
+            }
+
+            optimized_values.push_back(l);
+
+            if (globals.debug && l.limits_root_side.has_value()) {
+                EVLOG_info << "Sending enforced limits (import) to :" << l.uuid << " " << l.limits_root_side.value();
+            }
+        }
+    }
+
+    // Print out test case file
+    if (not test_name.empty()) {
+        json test_case;
+        test_case["start_time"] = Everest::Date::to_rfc3339(start_time);
+        test_case["request"] = json(request);
+        test_case["expected_result"] = json(optimized_values);
+        std::ofstream out(test_name.c_str());
+        out << test_case;
+        out.close();
+    }
+
+    return optimized_values;
+}
+
+} // namespace module
```

Comment:
missing newline, in general the new json files are missing a newline at the end as well